src/audio/frame.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <QDebug>
+#include <cmath>
+#include <limits>
+
+#include "engine/engine.h"
+
+namespace mixxx {
+namespace audio {
+/// FrameDiff_t can be used to store the difference in position between
+/// two frames and to store the length of a segment of track in terms of frames.
+typedef double FrameDiff_t;
+
+/// FramePos defines the position of a frame in a track
+/// with respect to a fixed origin, i.e. start of the track.
+class FramePos final {
+  public:
+    typedef double value_t;
+    static constexpr value_t kStartValue = 0;
+    static constexpr value_t kInvalidValue = std::numeric_limits<FramePos::value_t>::quiet_NaN();
+
+    constexpr FramePos()
+            : m_framePosition(kInvalidValue) {
+    }
+
+    constexpr explicit FramePos(value_t framePosition)
+            : m_framePosition(framePosition) {
+    }
+
+    static FramePos fromEngineSamplePos(double engineSamplePos) {
+        return FramePos(engineSamplePos / mixxx::kEngineChannelCount);
+    }
+
+    double toEngineSamplePos() const {
+        return value() * mixxx::kEngineChannelCount;
+    }
+
+    bool isValid() const {
+        return !std::isnan(m_framePosition);
```

Comment:
Does this actually work? I guess not. We probably need to use `util_isnan()` here. Do we have a test for this?

Reply:
I added a test, you're right.